be/src/storage/persistent_index.cpp

```diff
@@ -758,19 +782,72 @@ Status PersistentIndex::abort() {
+        std::string file_name = _get_l0_index_file_name(_path, _version);
+        // be maybe crash after create index file during last commit
+        // so we delete expired index file first to make sure no garbage left
+        Env::Default()->delete_file(file_name);
+        fs::BlockManager* block_mgr = fs::fs_util::block_manager();
+        std::unique_ptr<fs::WritableBlock> wblock;
+        fs::CreateBlockOptions wblock_opts({file_name});
+        RETURN_IF_ERROR(block_mgr->create_block(wblock_opts, &wblock));
+        DeferOp close_block([&wblock] {
+            if (wblock) {
+                wblock->close();
+            }
+        });
+        size_t snapshot_size = _dump_bound();
+        phmap::BinaryOutputArchive ar_out(file_name.data());
+        if (!_dump(ar_out)) {
+            std::string err_msg = strings::Substitute("faile to dump snapshot to file $0", file_name);
+            LOG(WARNING) << err_msg;
+            return Status::InternalError(err_msg);
+        }
+        // update PersistentIndexMetaPB
+        MutableIndexMetaPB* l0_meta = index_meta->mutable_l0_meta();
+        l0_meta->clear_wals();
+        IndexSnapshotMetaPB* snapshot = l0_meta->mutable_snapshot();
+        _version.to_pb(snapshot->mutable_version());
+        PagePointerPB* data = snapshot->mutable_data();
+        data->set_offset(0);
+        data->set_size(snapshot_size);
+        _offset += snapshot_size;
+        _page_size = 0;
+    } else {
+        MutableIndexMetaPB* l0_meta = index_meta->mutable_l0_meta();
+        IndexWalMetaPB* wal_pb = l0_meta->add_wals();
+        _version.to_pb(wal_pb->mutable_version());
 
-    PagePointerPB* data = wal_pb->mutable_data();
-    data->set_offset(_offset);
-    data->set_size(_page_size);
+        PagePointerPB* data = wal_pb->mutable_data();
+        data->set_offset(_offset);
+        data->set_size(_page_size);
 
-    _version.to_pb(index_meta->mutable_version());
-    index_meta->set_size(_size);
+        _version.to_pb(index_meta->mutable_version());
+        index_meta->set_size(_size);
+
+        _offset += _page_size;
+        _page_size = 0;
+    }
+    return Status::OK();
+}
+
+Status PersistentIndex::apply() {
+    if (_dump_snapshot) {
+        std::string expired_file_path = _index_block->path();
+        std::string index_file_path = _get_l0_index_file_name(_path, _version);
+        fs::BlockManager* block_mgr = fs::fs_util::block_manager();
+        std::unique_ptr<fs::WritableBlock> wblock;
+        fs::CreateBlockOptions wblock_opts({index_file_path});
+        // new index file should be created in commit() phase
+        wblock_opts.mode = Env::MUST_EXIST;
+        RETURN_IF_ERROR(block_mgr->create_block(wblock_opts, &wblock));
+        _index_block = std::move(wblock);
+        VLOG(1) << "delete expired l0 index file: " << expired_file_path;
+        Env::Default()->delete_file(expired_file_path);
+    }
+    _dump_snapshot = false;
```

Comment:
What happens to _dump_snapshot if commit failed or apply failed

Reply:
> What happens to _dump_snapshot if commit failed or apply failed

_dump_snapshot will be reset to `false` in the next time `prepare()` or `abort()` is called